File: swift-scraper/Sources/Models/Page.swift
//
//  Page.swift
//  swift-scraper
//
//  Created by Simon Liebers on 03.10.25.
//

import Foundation

struct Page: Codable {
    /// The original URL of the page
    let url: String

    /// The title of the page
    let title: String

    /// All links found on the page
    let links: [String]

    /// When the page was scraped
    let timestamp: Date

    /// Error message
    let error: String?

    // Custom initializer to create Page
    init(
        url: String,
        title: String = "",
        links: [String] = [],
        timestamp: Date = Date(),
        error: String? = nil
    ) {
        self.url = url
        self.title = title
        self.links = links
        self.timestamp = timestamp
        self.error = error
    }
}

File: go/models/page.go
package models

import (
	"time"
)

// Page represents the result of a web scraping operation.
// It contains both successful scraping results (title, links, images) and
// error information if the operation failed. A Page is always returned even
// on failure to maintain consistent result handling.
type Page struct {
	URL       string    `json:"url"`             // The original URL that was scraped
	Title     string    `json:"title"`           // The page title extracted from <title> tag
	Links     []string  `json:"links"`           // All href attributes from <a> elements
	Images    []string  `json:"images"`          // All src attributes from <img> elements
	TimeStamp time.Time `json:"timestamp"`       // When the scraping operation started
	Error     string    `json:"error,omitempty"` // Error message if scraping failed (empty on success)
}

// HasError reports whether the page scraping encountered an error.
// Returns true if the Error field is non-empty.
func (p *Page) HasError() bool {
	return p != nil && p.Error != ""
}

// Success reports whether the page was scraped successfully without errors.
// Returns true if the Error field is empty.
func (p *Page) Success() bool {
	return p != nil && p.Error == ""
}
